### PR TITLE
Added Solution to CF-786C

### PR DESCRIPTION
**Pull request by dustin-miao**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [ ] I have tested my code
- [ ] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [ ] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

**Comment by dutinmeow** [comment]
[Link to AC Code](https://codeforces.com/contest/786/submission/138572886)

File: solutions/cf-786C [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+A segment tree built over an array of size $n$ will visit at most $\mathcal{O}(\log n)$ nodes per query operation. Visiting each node takes $\mathcal{O}(1)$ time, for a total of $O(\log n)$ time. In our case, visiting each node takes $\mathcal{O}(n)$ time (to manually compute $f(m)$). Since we have $\mathcal{O}(\sqrt{n})$ "queries", each of which take $\mathcal{O}(n \log n)$ time to evaluate, our final complexity is $\mathcal{O}(n \sqrt{n} \log n)$. Although this may seem slow, the solution [passes comfortably within the time limit](https://codeforces.com/contest/786/submission/138572886). 
+
+<Info>
+Using `std::bitset` when evaluating $f$ will allow solutions with even slower time complexities to pass in time...
```
Comment (SansPapyrus683) [inline solutions/cf-786C:31]: don't add the 3 dots at the end, makes it sound weird

File: solutions/cf-786C [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+## Implementation
+
+<LanguageSection>
+
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int N, A[100005], ans[100005];
```
Comment (SansPapyrus683) [inline solutions/cf-786C:43]: you know the drill

File: solutions/cf-786C [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ 
+int f(int m) {
+	bitset<100005> B = 0;
```
Comment (SansPapyrus683) [inline solutions/cf-786C:46]: here too

File: solutions/cf-786C [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	int ret = 0, cnt = 0;
```
Comment (SansPapyrus683) [inline solutions/cf-786C:47]: assignments each on own line
also please use informative variable names, `ret` and `cnt` mean nothing

File: solutions/cf-786C [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+				B = 0;
+				cnt = 1;
+				B[A[i]] = 1;
+			}
+		}
+	}
+	return ret + !!cnt;
+}
+ 
+void dnq(int l = 1, int r = N, int lv = 1, int rv = N) {
```
Comment (SansPapyrus683) [inline solutions/cf-786C:63]:
```suggestion
void dnq(int l=1, int r=N, int lv=1, int rv=N) {
```

File: solutions/cf-786C [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	}
+	if (lv == rv) {
+		for (int i = l; i <= r; i++)
+			ans[i] = lv;
+		return;
+	}
+	int m = (l + r) / 2, mv = f(m);
+	ans[m] = mv;
+	dnq(l, m - 1, mv, rv);
+	dnq(m + 1, r, lv, mv);
```
Comment (SansPapyrus683) [inline solutions/cf-786C:75]: comments

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
*cracks knuckles*

**Comment by SansPapyrus683** [comment]
also if the editorial is awful dp then why did you put the problem in the persistent segtree
the least you could do is include the intended segtree solution as solution 1

**Comment by dutinmeow** [comment]
the intended solution is pst not dp?

**Comment by dutinmeow** [comment]
also the pst solution is horrible, I'm just saying that there's another way to do it that happens to have shorter code

**Comment by SansPapyrus683** [comment]
then... why did you put the problem in the pst part?

File: solutions/cf-786C (hunk 5 above)
Comment (dutinmeow) [reply solutions/cf-786C:63]: https://en.cppreference.com/w/cpp/language/default_arguments

Comment (dutinmeow) [reply solutions/cf-786C:63]: no need

**Comment by dutinmeow** [comment]
because it's also possible to solve it with PST

File: solutions/cf-786C [hunk 7]
```diff
@@ -40,11 +40,14 @@ Using `std::bitset` when evaluating $f$ will allow solutions with even slower ti
...
-int N, A[100005], ans[100005];
+const int MAXN = 1e5 + 1;
+
+int N, A[MAXN], ans[MAXN];
  
 int f(int m) {
-	bitset<100005> B = 0;
-	int ret = 0, cnt = 0;
+	bitset<MAXN> B = 0;
+	int ret = 0; // number of segments
```
Comment (SansPapyrus683) [inline solutions/cf-786C:49]: two spaces, you know...

**Comment by SansPapyrus683** [comment]
> because it's also possible to solve it with PST

then add the pst sol

File: solutions/cf-786C (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/cf-786C]: well of COURSE it means "set answer" and "recurse"
what does that recursion MEAN???

Comment (dutinmeow) [reply solutions/cf-786C]: you asked for it
its already descrbied in the problem statemnt

Comment (SansPapyrus683) [reply solutions/cf-786C]: bruh
also add comments for the if statement

**Comment by SansPapyrus683** [comment]
> well its in the official editorial why would i add something that already exists

because i looked at the official code and it's even worse than yours

File: solutions/cf-786C [hunk 8]
```diff
@@ -0,0 +1,95 @@
+---
+id: cf-786C
+source: CF
+title: Till I Collapse
+author: Dustin Miao
+---
+
+## Solution
+
+The [Official Editorial](https://codeforces.com/blog/entry/51163) contains a solution using persistant segment tree. There exists a much simpler, albeit slower, solution using divide and conquer.
```
Comment (SansPapyrus683) [inline solutions/cf-786C:10]: "persistant"

**Comment by SansPapyrus683** [comment]
dustin don't request a review when you haven't addressed all changes

**Comment by dutinmeow** [comment]
what is unaddressed?

**Comment by SansPapyrus683** [comment]
...you didn't add the pst sol

**Comment by dutinmeow** [comment]
I'm not addint the pst soln

**Comment by dutinmeow** [comment]
it's already in the editorial

**Comment by SansPapyrus683** [comment]
told you, the code was benq-tier

**Comment by dutinmeow** [comment]
doesn't matter, I'm still linking the editorial code
if you or someone else wants to transfer it over in a future pr, be my guest

**Comment by SansPapyrus683** [comment]
that's...not the problem, dustin.
the problem is that the code is awful, not that it isn't in the solution.

**Comment by dutinmeow** [comment]
like i said, it's not my problem

**Comment by dutinmeow** [comment]
i was just adding the dnq solution

**Comment by dutinmeow** [comment]
i don't want to

**Comment by SansPapyrus683** [comment]
alright, you're on your own

File: solutions/cf-786C [hunk 9]
```diff
@@ -0,0 +1,95 @@
+---
+id: cf-786C
+source: CF
+title: Till I Collapse
+author: Dustin Miao
+---
+
+## Solution
+
+The [Official Editorial](https://codeforces.com/blog/entry/51163) contains a solution using persistent segment tree. There exists a much simpler, albeit slower, solution using divide and conquer.
```
Comment (ryanchou-dev) [inline solutions/cf-786C:10]:
```suggestion
The [Official Editorial](https://codeforces.com/blog/entry/51163) contains a solution using persistent segment tree. However, there exists a much simpler, albeit slower, solution using divide and conquer. 
```

File: solutions/cf-786C [hunk 10]
```diff
@@ -0,0 +1,95 @@
+---
+id: cf-786C
+source: CF
+title: Till I Collapse
+author: Dustin Miao
+---
+
+## Solution
```
Comment (ryanchou-dev) [inline solutions/cf-786C:8]: maybe add an `Intended Solution` with a `see official analysis` and then `Alternate Solution - DNQ` for this one

**Review by ryanchou-dev — APPROVED** [review]

**Review by nwatx — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
lol

File: solutions/cf-786C (hunk 2 above)
Comment (DenialRiver8810) [reply solutions/cf-786C]: do we have to just do 100000 in the solutions?

Comment (SansPapyrus683) [reply solutions/cf-786C]: no it’s just something i want

**Pull request closed without merging**